### PR TITLE
Updated to the latest beta of lsp.  Spec version 3.16

### PR DESCRIPTION
**Pull request by david-driscoll**
The latest version of the spec just dropped, this beta has all the latest spec features, plus a few minor breaking changes, documented with the release [here](https://github.com/OmniSharp/csharp-language-server-protocol/releases/tag/v0.19.0-beta.4)

### CODE REVIEW

**Comment by david-driscoll** [comment]
NOTE: I'll probably remove this from beta in about 2 weeks, everything should be good, unless there is feedback that folks are unhappy with the minor breaking changes.

File: build/Settings.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
   <PropertyGroup>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>9.0</LangVersion>
```
Comment (david-driscoll) [inline build/Settings.props:5]: @filipw @JoeRobich anyone have an issue with this change?

Comment (JoeRobich) [reply build/Settings.props:5]: I don't have a problem with it.

Comment (david-driscoll) [reply build/Settings.props:5]: good because the latest library has lots of records... lots of with statements 🤣

Comment (filipw) [reply build/Settings.props:5]: one small issue with this is that some features of C# 9 e.g. positional records or `init` are not available without .NET 5.0, so it is easy to fall into the trap of using something that won't compile 😄

**Review by filipw — APPROVED** [review]
looks good to me

File: build/Settings.props (hunk 1 above)
Comment (david-driscoll) [reply build/Settings.props:5]: records have been working fine for me, the only gotcha is much like NRT you have to provide an interop type.  
https://github.com/OmniSharp/csharp-language-server-protocol/blob/81a38cca80a4e35ec198de75dbab5a5571aae82a/src/JsonRpc/Nullable/Records.cs

**Comment by razzmatazz** [comment]
I was trying to figure out the cause for #1995 with the latest changes checked out from this PR but still failed -- the issue contains updated stack trace.

Sorry for trying to shamelessly plug in another issue, but hopefully this is looks to be related/easy to fix?

This would allow me to implement other features which are now available only via native omnisharp protocol, like assembly inspection/metadata/test runners, etc in `lsp-mode` on emacs and in other editors too.

**Comment by razzmatazz** [comment]
> Sorry for trying to shamelessly plug in another issue, but hopefully this is looks to be related/easy to fix?
> 
> I was trying to figure out the cause for #1995 with the latest changes checked out from this PR but still failed -- the issue contains updated stack trace.
> 
> This would allow me to implement other features which are now available only via native omnisharp protocol, like assembly inspection/metadata/test runners, etc in `lsp-mode` on emacs and in other editors too.

I have created a PR to OS.E.LSP that fixes my issue:
 - https://github.com/OmniSharp/csharp-language-server-protocol/pull/488

So I guess my previous comment with a shameless plug can be ignored here.

**Review by JoeRobich — APPROVED** [review]

**Comment by razzmatazz** [comment]
I can confirm this branch works works (for me, at least).

With [this patch in O.E.L](https://github.com/OmniSharp/csharp-language-server-protocol/commit/32a135922dcf794158588fdacd66213b102ea5c4) it enables me to use the legacy endpoints via LSP to enable test-runner, metadata browsing and other functionality in emacs, at least. This will also allow us to use the `codeaction/resolve` method from LSP 3.16 and drop the custom `omnisharp/executeCodeAction` method as per #2068.

Would it be feasible to merge this PR before the next release? This unlocks quite a few things LSP-wise.

File: src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCodeActionHandler.cs [hunk 2]
```diff
@@ -101,14 +95,14 @@ public override async Task<CommandOrCodeActionContainer> Handle(CodeActionParams
...
                         Command = Command.Create("omnisharp/executeCodeAction")
-                            .WithTitle(ca.Name)
                             .WithArguments(new CommandData()
                             {
                                 Uri = request.TextDocument.Uri,
                                 Identifier = ca.Identifier,
                                 Name = ca.Name,
                                 Range = request.Range,
                             })
+                            with { Title = ca.Name }
```
Comment (333fred) [inline src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCodeActionHandler.cs:105]: We can't also with the arguments?

File: src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@
 
 namespace OmniSharp.LanguageServerProtocol.Handlers
 {
-    class OmniSharpCompletionHandler : CompletionHandler
+    class OmniSharpCompletionHandler : CompletionHandlerBase
```
Comment (333fred) [inline src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs:14]: Man, this whole handler needs to be ported to the new one, doesn't it.

File: build/Settings.props (hunk 1 above)
Comment (333fred) [reply build/Settings.props:5]: I've tried to use `or` patterns several times now, please give them to me :)

File: src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCodeActionHandler.cs (hunk 2 above)
Comment (david-driscoll) [reply src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCodeActionHandler.cs:105]: as it stands yes and no..., the arguments is a [JArray](https://github.com/OmniSharp/csharp-language-server-protocol/blob/e0c9fc66da91ab389874f34ff9a5b1b18d932be6/src/Protocol/Models/Command.cs#L27-L28) so you'd have to do `with { Arguments = JArray.FromObject(...) }`

**Review by 333fred — APPROVED** [review]

File: src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs (hunk 3 above)
Comment (filipw) [reply src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs:14]: yes

**Review by filipw — APPROVED** [review]
LGTM

File: src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs (hunk 3 above)
Comment (david-driscoll) [reply src/OmniSharp.LanguageServerProtocol/Handlers/OmniSharpCompletionHandler.cs:14]: thankfully the changed merged in cleanly.

**Pull request merged**